contracts/TransferLimitModule.sol

```diff
@@ -0,0 +1,417 @@
+        }
+    }
+
+    /// @dev Allows to update the transfer limit for a specified token. This can only be done via a Safe transaction.
+    /// @param token Token contract address.
+    /// @param transferLimit Transfer limit in smallest token unit.
+    function changeTransferLimit(address token, uint256 transferLimit)
+        public
+        authorized
+    {
+        transferLimits[token].transferLimit = transferLimit;
+    }
+
+    /// @dev Updates the delegate. This can only be done via a Safe transaction.
+    /// @param _delegate New delegate, who is a non-owner account also allowed to make transfers.
+    function setDelegate(address _delegate)
+        public
+        authorized
+    {
+        delegate = _delegate;
+    }
+
+    /// @dev Returns if Safe transaction is a valid transfer limit transaction.
+    /// @param token Address of the token that should be transfered (0 for Ether)
+    /// @param to Address to which the tokens should be transfered
+    /// @param amount Amount of tokens (or Wei) that should be transfered
+    /// @param safeTxGas Gas that should be used for the Safe transaction.
+    /// @param dataGas Gas costs for data used to trigger the safe transaction and to pay the payment transfer
+    /// @param gasPrice Gas price that should be used for the payment calculation.
+    /// @param gasToken Token address (or 0 if ETH) that is used for the payment.
+    /// @param refundReceiver Address of receiver of gas payment (or 0 if tx.origin).
+    /// @param signatures Packed signature data ({bytes32 r}{bytes32 s}{uint8 v})
+    /// @return Returns if transaction can be executed.
+    function executeTransferLimit(
+        address token,
+        address to,
+        uint256 amount,
+        uint256 safeTxGas,
+        uint256 dataGas,
+        uint256 gasPrice,
+        address gasToken,
+        address refundReceiver,
+        bytes signatures
+    )
+        public
+    {
+        require(to != 0, "Invalid to address provided");
+        require(amount > 0, "Invalid amount provided");
+
+        uint256 startGas = gasleft();
+        bytes32 txHash = getTransactionHash(
+            token, to, amount,
+            safeTxGas, dataGas, gasPrice, gasToken, refundReceiver,
+            nonce
+        );
+        require(checkSignatures(txHash, signatures), "Invalid signatures provided");
+        // Increase nonce and execute transaction.
+        nonce++;
+        require(gasleft() >= safeTxGas, "Not enough gas to execute safe transaction");
+
+        // Validate that transfer is not exceeding transfer limit, and
+        // update state to keep track of spent values.
+        require(handleTransferLimits(token, amount), "Transfer exceeds limits");
+
+        // Perform transfer by invoking manager
+        if (token == 0) {
+            require(manager.execTransactionFromModule(to, amount, "", Enum.Operation.Call), "Could not execute ether transfer");
```

Comment:
In the safe don't revert if the user transaction fails. Would be interesting if that makes sense here too (same for `handleTransferLimits`)

Lets assume somebody submits a transferlimit request but at the same time somebody else transfers out the last tokens from the Safe. Who should be held responsible and cover the gas costs?

Reply:
Hmm, as it is now it's the relayer who bears the cost when the user transaction fails, which doesn't seem right. I'll modify it so as not to revert on user transaction failure, so that relayer can still get refunded (if any fund is in the safe!).

Regarding `handleTransferLimits`, I went with reverting because apart from validation it also modifies the state (spent values).

Reply:
I just realized that if the module doesn't revert on transfer failure, the spent values could get updated in `handleTransferLimits`, without any actual transfer taking place! (for example when there's not enough balance in safe)

Should I check limits before transfer, and only update spent values after?